stm-doc/index.mld

```diff
@@ -0,0 +1,451 @@
+    mutable length  : int }
+
+  let empty () = { content = []; length = 0 }
+
+  let mem a t = List.mem a t.content
+
+  let length t = t.length
+
+  let add a t =
+    if not (mem a t)
+    then begin
+      t.content <- a :: t.content;
+    end
+end
+]}
+
+The first thing to do is to provide a description of this library. We call this
+description a specification and it takes the form of a module of type
+{!module-STM_base.Spec}. In this module, the user describes three things:
+{ol
+{- the library to be tested,}
+{- the functional model against which to test it,}
+{- and how to generate commands from this library.}
+}
+
+The model should be simple enough to be obviously correct. The important thing
+to keep in mind is that the model doesn't have to have exactly the same
+interface as the library to be tested. So no need to have a mutable set as
+model, we can use the immutable {!module-Stdlib.Set} library, or build one if
+we prefer. Here, we will use the one from the standard library.
+
+We should also choose how to instantiate the type parameter ([Lib.t] is a
+container, but if we want to test it, we should precise what type of things it
+contains). Let's say it contains [int]s.
+
+{[
+module Lib_spec : Spec = struct
+
+  module State = Set.Make (struct type t = int let compare = Int.compare end)
+
+  type sut        = int Lib.t
+  let init_sut () = Lib.empty ()
+  let cleanup _   = ()
+
+  type cmd =
+    | Mem of int
+    | Add of int
+    | Length [@@deriving show { with_path = false }]
+
+  type state     = State.t
+  let init_state = State.empty
+
+  let arb_cmd _state =
+    QCheck.make ~print:show_cmd
+      (QCheck.Gen.oneof
+        [Gen.return Length;
+         Gen.map (fun i -> Mem i) Gen.int;
+         Gen.map (fun i -> Add i) Gen.int;
+        ])
+
+  let next_state cmd state =
+    match cmd with
+    | Mem _  -> state
+    | Add i  -> State.add i state
+    | Length -> state
+
+  let run cmd sut =
+    match cmd with
+    | Mem i  -> Res (bool, Lib.mem i sut)
+    | Add i  -> Res (unit, Lib.add i sut)
+    | Length -> Res (int, Lib.length sut)
+
+  let precond _ _ = true
+
+  let postcond cmd state res =
+    match cmd, res with
+    | Mem i,  Res ((Bool,_), b) -> b = State.mem i state
+    | Length, Res ((Int,_), l)  -> l = State.cardinal state
+    | Add _,  Res ((Unit,_),_)  -> true
+    | _                           -> false
+end
+]}
+
+In [arb_cmd]'s definition, note that [QCheck.make] takes an optional printer.
+It is important to provide it so that the test's output is readable. Here, we
+use a [ppx_deriver], but you can provide your own printers if you want to.
+
+Another remark worth making, is that the [state] argument in the [postcond]
+function is the state of the model {i before} running the command (that is
+before computing [next_state cmd state]). This will be important when we will
+add a {{!section-v5}[remove]} function to our set library.
+
+Now, we are set to run our first [qcheck-stm] tests!
```

Comment:
I think it is worth taking the reader by the hand here through the above definitions, ideally explaining how points 1,2,3 above are expressed with them.